### PR TITLE
EOS-24424: cortx_setup cluster bootstrap

### PR DESCRIPTION
**Pull request by selvakumaar5496**
Signed-off-by: Selvakumar <selva.k.nambi@seagate.com>

# Problem Statement
- cortx_setup cluster bootstrap
  Invoke MiniProvisioner interfaces such as post_install, prepare, config, init

# Design
-  https://jts.seagate.com/browse/EOS-24424

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

File: src/provisioner/provisioner.py [hunk 1]
```diff
@@ -96,11 +96,17 @@ def cluster_bootstrap(node_id: str, cortx_conf_url: str = None):
...
-            services = cortx_config_store.get(f'node>{node_id}>components[{j}]>services')
-            service = 'all' if services is None else ','.join(services)
-            print(f"{components[j]['name']}_setup --config {CortxProvisioner._cortx_conf_url} --services %s" %service)
-
-            # TODO: Enable this code
-            # rc, output = SimpleProcess(f"{components[i]}_setup --config \
-            #  {CortxProvisioner._cortx_conf_url} --services %s" %services.join(","))
+        mp_interfaces = ['post_install', 'prepare', 'config', 'init']
+        for interface in mp_interfaces:
+            for j in range(0, num_components):
```
Comment (ujjwalpl) [inline src/provisioner/provisioner.py:101]: Orchestration goes like this
1. comp1:post_install
2. comp2: post_install
3. comp1:prepare
4. comp2:prepare..
5. ...

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/provisioner/provisioner.py (hunk 1 above)
Comment (selvakumaar5496) [reply src/provisioner/provisioner.py:101]: yes @ujjwalpl 

Below is the flow
```
motr_setup --config yaml:///tmp/services.yaml post_install --services io
hare_setup --config yaml:///tmp/services.yaml post_install --services all
s3_setup --config yaml:///tmp/services.yaml post_install --services io,auth,bg_consumer
utils_setup --config yaml:///tmp/services.yaml post_install --services message_bus
motr_setup --config yaml:///tmp/services.yaml prepare --services io
hare_setup --config yaml:///tmp/services.yaml prepare --services all
s3_setup --config yaml:///tmp/services.yaml prepare --services io,auth,bg_consumer
utils_setup --config yaml:///tmp/services.yaml prepare --services message_bus
motr_setup --config yaml:///tmp/services.yaml config --services io
hare_setup --config yaml:///tmp/services.yaml config --services all
s3_setup --config yaml:///tmp/services.yaml config --services io,auth,bg_consumer
utils_setup --config yaml:///tmp/services.yaml config --services message_bus
motr_setup --config yaml:///tmp/services.yaml init --services io
hare_setup --config yaml:///tmp/services.yaml init --services all
s3_setup --config yaml:///tmp/services.yaml init --services io,auth,bg_consumer
utils_setup --config yaml:///tmp/services.yaml init --services message_bus
```

File: src/setup/cortx_setup.py [hunk 2]
```diff
@@ -66,15 +68,17 @@ def add_args(parser: str):
 
     def process(self, *args, **kwargs):
         """ Bootsrap Cluster """
-        if self._args.action == "bootstrap":
-            CortxProvisioner.cluster_bootstrap(self._args.node_id, self._args.cortx_conf)
+        if self._args.action == 'bootstrap':
```
Comment (ujjwalpl) [inline src/setup/cortx_setup.py:71]: For consistency with config apply, lets add -f command option.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_setup.py (hunk 2 above)
Comment (selvakumaar5496) [reply src/setup/cortx_setup.py:71]: Done

File: src/provisioner/provisioner.py [hunk 3]
```diff
@@ -140,12 +139,24 @@ def cluster_bootstrap(node_id: str, cortx_conf_url: str = None):
...
-            # rc, output = SimpleProcess(f"{components[i]}_setup --config \
-            #  {CortxProvisioner._cortx_conf_url} --services %s" %services.join(","))
+        num_components = len(components)
+        mp_interfaces = ['post_install', 'prepare', 'config', 'init']
+        for interface in mp_interfaces:
+            for j in range(0, num_components):
+                services = cortx_config_store.get(
+                    f'node>{node_id}>components[{j}]>services')
+                service = 'all' if services is None else ','.join(services)
+                cmd_proc = SimpleProcess(f"{components[j]['name']}_setup " \
```
Comment (ujjwalpl) [inline src/provisioner/provisioner.py:155]: Rebase your code. Old code is not the current code.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by ujjwalpl — APPROVED** [review]

**Review by sachinpunadikar — APPROVED** [review]
+1

**Pull request merged**